app/models/akka/RequesterBehavior.scala

```diff
@@ -45,10 +125,40 @@ trait RequesterBehavior { me: AbstractDSLinkActor =>
   def stopRequester() = {
     batchAndRoute(targetsByRid map { case (rid, target) => target -> CloseRequest(rid) })
     batchAndRoute(targetsBySid.zipWithIndex map {
-      case ((sid, target), index) => target -> UnsubscribeRequest(lastRid + index + 1, sid)
+      case ((sid, PathAndQos(target, _)), index) => target -> UnsubscribeRequest(lastRid + index + 1, sid)
     })
   }
 
+  private def handleSubscriptions(subscriptions:Seq[DSAResponse], connected: Boolean = true) = subscriptions foreach {
+    r => withQosAndSid(r) foreach {
+      message =>
+      log.debug(s"sending subscription message: $message")
+      val toSend = SubscriptionNotificationMessage(-1, None, List(message.response), message.sid, message.qos)
+
+      if(connected){
+        // in connected state pushing to stream with backpressure logic
+        toSocket offer toSend
+      } else if(message.qos >= QoS.Durable){
+        //in disconnected - just send to state actor
+        qosState ! PutNotification(toSend)
+      }
+    }
+  }
+
+  private def isSubscription(response:DSAResponse):Boolean = response.rid == 0
+
+  private def withQosAndSid(response:DSAResponse):Seq[SubscriptionResponseEnvelope] = {
+    response.updates.map{
+      updates =>
+        updates.map{
+          update =>
+            val sid = extractSid(update)
+            val qos = targetsBySid.get(sid).map(_.qos) getOrElse(QoS.Default)
+            SubscriptionResponseEnvelope(response, sid, qos)
+        }
+    } getOrElse(List())
+  }
```

Comment:
when i see two nested `=>`, my hand reaches for `for {...}`
but that's just me bitching... ))

Reply:
It's Option[Seq], so i don't think for comprehention would work